st.experimental_audio_input

sfc-gh-nbellante opened the pull request:
## Describe your changes

🔉 This PR introduces the `st.experimental_audio_input` widget! 🎤 

### Light Theme Demo
![Kapture 2024-09-13 at 11 44 38](https://github.com/user-attachments/assets/56530c43-31f4-498c-b322-df52f6be804a)


### Dark Theme Demo
![Kapture 2024-09-13 at 11 48 59](https://github.com/user-attachments/assets/70069859-bc58-4eb8-ad7c-8aff6bdaf74f)


### Light Theme No Mic Permissions
<img width="755" alt="image" src="https://github.com/user-attachments/assets/51df40c7-1c75-415b-bc70-59fdb59f1c3b">

### Dark Theme No Mic Permissions
<img width="752" alt="image" src="https://github.com/user-attachments/assets/8f8d47f6-2195-4dfc-ab41-07231ae3a15b">

Note: this currently does not work on safari and by extension iOS so users on that platform will see a warning with the audio_input disabled
![image](https://github.com/user-attachments/assets/0b882904-3f66-464e-9a9b-3587805a33a9)



## GitHub Issue Link (if applicable)
https://github.com/streamlit/streamlit/issues/8899
https://github.com/streamlit/streamlit/issues/903

## Testing Plan

- Unit Tests (JS and/or Python)
   - python unit tests added ✅ 
- E2E Tests
  - e2e tests added ✅ 
  - Note: was unable to reliably add invalid permissions check test that would work locally
- Any manual testing needed?
  - was also tested manually and will be QA'd

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
why isnt it working ;(

sfc-gh-nbellante committed:
whoops adding these files too

sfc-gh-nbellante committed:
getting it working

sfc-gh-nbellante committed:
clean it up a lil

kajarenc committed:
Adding functionality for st.audio_input on python side (#9111)

sfc-gh-nbellante committed:
basic ui in place

sfc-gh-nbellante committed:
this file too

sfc-gh-nbellante committed:
more cleanup and actually working e2e

sfc-gh-nbellante committed:
adding some prelim tests and making build pass

sfc-gh-nbellante committed:
commenting out debug statement

sfc-gh-nbellante committed:
removing react-media-recorder

sfc-gh-nbellante committed:
make it work on safari

sfc-gh-nbellante committed:
test

sfc-gh-nbellante committed:
more test

sfc-gh-nbellante committed:
fix type

sfc-gh-nbellante committed:
attempt to fix build and add adj height 2fer

sfc-gh-nbellante committed:
fix

sfc-gh-nbellante committed:
minor visual tweaks to the hacky controls

sfc-gh-nbellante committed:
try to force prompt for perms

sfc-gh-nbellante committed:
bunch of cleanup and adhering more to the design

sfc-gh-nbellante committed:
adjusting dots

sfc-gh-nbellante committed:
dark mode colors 99%

sfc-gh-nbellante committed:
fix going to play when ends

sfc-gh-nbellante committed:
timer playback matching design

sfc-gh-nbellante committed:
fixing counter text color

sfc-gh-nbellante committed:
no mic perms

sfc-gh-nbellante committed:
add some comments and adjust button disablement styling

sfc-gh-nbellante committed:
refactor placeholder

sfc-gh-nbellante committed:
cleanup placeholder more

sfc-gh-nbellante committed:
removing media stream visualizer

sfc-gh-nbellante committed:
deleting unused dependencies

sfc-gh-nbellante committed:
more unused files

sfc-gh-nbellante committed:
refactoring formatTime util

sfc-gh-nbellante committed:
more cleanup and getting label and toolbar working nicely together

sfc-gh-nbellante committed:
fixing some colors

sfc-gh-nbellante committed:
revoking the url upon clearing

sfc-gh-nbellante committed:
more cleanup, more refactor, more optimized

sfc-gh-nbellante committed:
removing another non-necessary file

sfc-gh-nbellante committed:
fix python

sfc-gh-nbellante committed:
adding more extensive e2e tests on frontend

sfc-gh-nbellante committed:
updating height

sfc-gh-nbellante committed:
fixing the delete file bug

kajarenc committed:
Nico/audio input widget protoype backend polishing (#9399)

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: Nico Bellante <nico.bellante@snowflake.com>

sfc-gh-nbellante committed:
lint

sfc-gh-nbellante committed:
more lint

sfc-gh-nbellante committed:
update notices

sfc-gh-nbellante committed:
maybe fix test

sfc-gh-nbellante committed:
lets try this!

sfc-gh-nbellante committed:
third times the charm

sfc-gh-nbellante committed:
oh lol gotta love tests, i think it actually caught a bug introduced when the linter yelled at me about hook deps

sfc-gh-nbellante committed:
ughhhhhhhhh

sfc-gh-nbellante committed:
another hook deps error

sfc-gh-nbellante committed:
fix

sfc-gh-nbellante committed:
okay, just chomium, you win playwright

sfc-gh-nbellante committed:
one more hook eslint ignore

sfc-gh-nbellante committed:
adding disabled

sfc-gh-nbellante committed:
cleaning up styling

sfc-gh-nbellante committed:
updating padding

sfc-gh-nbellante committed:
fix popover test

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,339 @@
+  )
+
+  return (
+    <StyledAudioInputContainerDiv>
```
nit: the top-level wrapper needs a `stAudioInput` class (for the improved CSS support) and for consitency it is also preferred to have the test ID on the top level.

sfc-gh-nbellante replied:
Ah so we add these classes so that people can customize them later?

lukasmasuch replied:
Yep, the only reason why the classes exist is to provide a slightly more stable/convenient way for users to apply CSS hacks to our components. It doesn't have any other purpose.

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,339 @@
+        </StyledWaveformInnerDiv>
+        <StyledWaveformTimeCode
+          isPlayingOrRecording={isPlayingOrRecording}
+          data-testid="StyledWaveformTimeCode"
```
nit: for consistency and to prevent unwanted overlap, we have a naming schema defined for test-ids (and class-names) which is `stComponentNameSubComponent`, in this case something like: `stAudioInputWaveformTimeCode`

sfc-gh-nbellante replied:
lol of course i happen to follow the one instance that didn't follow this pattern. will update
<img width="565" alt="image" src="https://github.com/user-attachments/assets/298d873c-fd12-4970-b9ea-2a01e4ca5fe1">

lukasmasuch replied:
:) oh, yeah... I ignored the `StyledFullScreenButton` with the recent cleanups since it is anyways planned to be fully removed (replaced by toolbar)

sfc-gh-bnisco commented on frontend/lib/src/components/core/Block/ElementNodeRenderer.tsx:
```diff
@@ -497,6 +502,20 @@ const RawElementNodeRenderer = (
 
+    case "audioInput": {
+      const audioInputProto = node.element.audioInput as AudioInputProto
+      widgetProps.disabled = widgetProps.disabled || audioInputProto.disabled
```
**thought:** It is counterintuitive to me that we mutate `widgetProps`, although I see us doing this in many other places in this file.

My first reaction was that we should not mutate props, but then I realized that `widgetProps` itself is an object defined locally in this function. With that said, we do this in a lot of places throughout this file. I wonder if there's a better abstraction we could build in to prevent us from having to duplicate this logic everywhere. Note that this not something we should do in this PR.

sfc-gh-nbellante replied:
💯 I found this a bit strange too, let's put this in the backlog!

sfc-gh-nbellante replied:
added it to the Tech Debt DB

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,339 @@
+  )
+}
+
+export default withTheme(AudioInput)
```
nit: I think for functional components the `useTheme` hook is preferred. The `withTheme` is more of a historical relict used by class-based components:

```typescript
const theme: EmotionTheme = useTheme()
```

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/ActionButton.tsx:
```diff
@@ -0,0 +1,112 @@
+}) => {
+  const theme = useTheme()
+
+  if (recordPlugin && recordPlugin.isRecording()) {
```
**nit:** Optional chaining is often nicer 

```suggestion
  if (recordPlugin?.isRecording()) {
```

lukasmasuch commented on lib/streamlit/elements/widgets/audio_input.py:
```diff
@@ -0,0 +1,248 @@
+        args: WidgetArgs | None = None,
+        kwargs: WidgetKwargs | None = None,
+        *,  # keyword-only arguments:
+        disabled: bool = False,
```
for all the new elements we try to only have the absolute minimum as positional parameters, which most likely would be only `label` here.

sfc-gh-nbellante replied:
@kajarenc can you make these changes?

kajarenc replied:
Done!

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/ActionButton.tsx:
```diff
@@ -0,0 +1,112 @@
+      kind={BaseButtonKind.BORDERLESS_ICON}
+      onClick={startRecording}
+      disabled={hasNoMicPermissions || disabled}
+      data-testid="stAudioInputRecordButton"
```
**suggestion:** Rather than using a `data-testid` to be able to access these elements, we should prefer to utilize accessible methods instead.

RTL is opinionated on how to select elements, with a [clear priority order](https://testing-library.com/docs/queries/about#priority). Giving this an [aria-label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) rather than a testId is both a win for accessibility and for testing! Note that this comment here applies to all the buttons in this component.

sfc-gh-nbellante replied:
Good call, let's do it!

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/NoMicPermissions.tsx:
```diff
@@ -0,0 +1,40 @@
+
+const NoMicPermissions = (): ReactElement => {
+  return (
+    <StyledNoMicInputContainerDiv data-testid="StyledNoMicInputContainerDiv">
```
**suggestion:** Similar to the above, this `testid` shouldn't be necessary. Instead the tests should get elements via accessible manners, such as `getByText`

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/formatTime.ts:
```diff
@@ -0,0 +1,28 @@
+ * limitations under the License.
+ */
+
+const formatTime = (timeMs: number): string => {
```
**question:** Any particular reason we are writing our own formatter instead of leveraging something built in like `date.toLocaleTimeString('en-US', { minute: '2-digit', second: '2-digit' })`?

Now I know that this example includes a hardcoded `en-US`, but I suppose that brings up a follow-up question, should we consider supporting more locales out of the box? If we wanted to utilize `undefined` instead of `en-US`, we should get locale support for free.

sfc-gh-nbellante replied:
fantastic idea!

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/styled-components.ts:
```diff
@@ -0,0 +1,83 @@
+  isPlayingOrRecording: boolean
+}>(({ theme, isPlayingOrRecording }) => ({
+  margin: 8,
+  fontFamily: "Source Code Pro, monospace",
```
**suggestion:** Should be using the existing `fonts.monospace`

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/styled-components.ts:
```diff
@@ -0,0 +1,83 @@
+export const StyledWaveformTimeCode = styled.span<{
+  isPlayingOrRecording: boolean
+}>(({ theme, isPlayingOrRecording }) => ({
+  margin: 8,
```
**suggestion:** Should be using the existing `spacing` scale (this is true for everything in this file)

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/styled-components.ts:
```diff
@@ -0,0 +1,83 @@
+    ? theme.genericColors.bodyText
+    : theme.colors.fadedText60,
+  backgroundColor: theme.genericColors.secondaryBg,
+  fontSize: 14,
```
**suggestion:** Should be using rem-based font size from `fontSizes`

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/uploadFiles.ts:
```diff
@@ -0,0 +1,113 @@
+  try {
+    fileUrls = await uploadClient.fetchFileURLs(files)
+  } catch (e) {
+    console.error("Error fetching file URLs")
```
**question:** Do we typically keep these kinds of logs in?

sfc-gh-nbellante replied:
Nah that was just missed

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,339 @@
+        }
+      })
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
**question:** Why are we disabling this? Will it lead to stale values in the callback closure?

sfc-gh-nbellante replied:
I didn't get enough time to look into _why_ this is happening, but it seems that element is changing much more than I expected. Luckily, we're just using the element for its [WidgetInfo](https://github.com/streamlit/streamlit/blob/nico/audio-input/frontend/lib/src/WidgetStateManager.ts#L43) so we don't really care about those updates. 

We could probably clean this up by extracting that widget info and including that new object in the deps array? wdyt?

Oh yeah and to why this was a problem: the `intitializeWavesurfer` depends on this function, and it would cause that to re-run and that re-running causes wavesurfer to be reinitialized and loses all the local state. 😞

sfc-gh-nbellante replied:
> Will it lead to stale values in the callback closure?

No, but I think we can make this more self documenting.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,339 @@
+    return () => {
+      if (ws) ws.destroy()
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
**question:** Why are we disabling this? Will it lead to stale values in the callback closure?

sfc-gh-nbellante replied:
This was initially a quick fix for a linter issue, but the code in question is no longer needed. I’ll remove both the line and the disable. Thanks for catching that!

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    wavesurfer.setOptions({
+      waveColor: theme.colors.primary,
```
**question:** Is it intentional that this one uses `theme.colors`, but the one above uses a hardcoded hex?

sfc-gh-nbellante replied:
Yes, and for a bit of a frustrating reason. (see [comment](https://github.com/streamlit/streamlit/pull/9404/files#diff-a7eb8abbe0fa76f9716f9f399ee823661f1a6c761fa6065ef32204746da10ad4R175)) tldr: the way you'd expect wavesurfer to be using the colors is not how it's using the colors and the alpha is getting all messed up and it's like applying the progress color as a tint instead of just making the color that color

sfc-gh-nbellante replied:
going to add this to the tech debt db too just so we don't lose track of it

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,339 @@
+      !hasNoMicPermissions
+  )
+
+  const showNoMicPermissionsOrPlaceholder = useMemo(
+    () => Boolean(hasNoMicPermissions || showPlaceholder),
+    [hasNoMicPermissions, showPlaceholder]
+  )
```
**suggestion (non-blocking):** I think it'd be fine to not wrap this in a `useMemo`, firstly because it matches the above boolean values, but also more broadly speaking because a `useMemo` is probably more overhead than necessary for this simple boolean comparison

sfc-gh-nbellante replied:
fair enough

sfc-gh-nbellante committed:
Update frontend/lib/src/components/widgets/AudioInput/ActionButton.tsx

Co-authored-by: Bob Nisco <bob.nisco@snowflake.com>

sfc-gh-nbellante committed:
removing deps ignore

sfc-gh-nbellante committed:
removing console logs

sfc-gh-nbellante committed:
monospace theme font

sfc-gh-nbellante committed:
i knew i forgot something lol

sfc-gh-nbellante committed:
leverage js date formatter for time formatting the counter

sfc-gh-nbellante committed:
removing memo

sfc-gh-nbellante committed:
no more memo

sfc-gh-nbellante committed:
removing ignore deps lint comment

sfc-gh-nbellante committed:
a win for a11y is a win for a11

raethlein commented on e2e_playwright/st_audio_input_test.py:
I think we should include a snapshot test here using a themed app to ensure the audio widget's look in light and dark theme and across changes

sfc-gh-nbellante replied:
good call! i'll add that

raethlein commented on e2e_playwright/st_fragment_basics.py:
```diff
@@ -23,9 +23,9 @@
-# return values. We also don't test the camera_input, data_editor, and file_uploader
-# widgets as well as custom components here due to the disproportionate amount of work
-# required to do so.
+# return values. We also don't test the audio_input, camera_input, data_editor, and
```
just for documentation purposes, @kajarenc and I tested it manually yesterday. It's probably a good idea to verify this and also state this explicitly in the PR description 🙂

sfc-gh-nbellante replied:
Wait, so what's the ask here? To confirm that `st.audio_input` doesn't work with fragments?

raethlein replied:
It should work in fragments 🙂 Sorry for the unclear message; I meant that it'd be great to state in the PR description that `st.audio_input` has been manually tested with fragments.

sfc-gh-nbellante replied:
it now works in fragments!

raethlein commented on frontend/lib/src/components/shared/Toolbar/Toolbar.tsx:
```diff
@@ -40,17 +40,15 @@ export interface ToolbarActionProps {
-  onClick,
-}: ToolbarActionProps): ReactElement {
+export function ToolbarAction(props: ToolbarActionProps): ReactElement {
+  // eslint-disable-next-line @typescript-eslint/naming-convention
```
nit: instead of silencing the warning, would it make sense to rename the props (likely `show_label`, right?) or is this beyond our control?
If we cannot rename the incoming props, would it remove the warning if we renamed it during restructuring like `const { label, show_label: showLabel, icon, onClick } = props`?

sfc-gh-nbellante replied:
i think i can actually revert this change completely since i no longer need the test id here

raethlein commented on frontend/lib/src/components/shared/Toolbar/Toolbar.tsx:
```diff
@@ -40,17 +40,15 @@ export interface ToolbarActionProps {
   const displayLabel = show_label ? label : ""
   return (
-    <div data-testid="stElementToolbarButton">
+    <div data-testid={"stElementToolbarButton"}>
```
nit: I think for strings you don't need the `{ }` brackets. Is this something a linter or so highlighted? In this case, I would vote for filing a tech debt item and do it consistently across the code base

sfc-gh-nbellante replied:
nah, i'll revert this. it's because it used to have a `props["data-testid"] ??` sitting in it

raethlein commented on frontend/lib/src/components/widgets/AudioInput/ActionButton.tsx:
```diff
@@ -0,0 +1,112 @@
+  if (recordPlugin?.isRecording()) {
+    // It's currently recording, so show the stop recording button
+    return (
+      <BaseButton
+        kind={BaseButtonKind.BORDERLESS_ICON}
+        onClick={stopRecording}
+        disabled={disabled}
+        aria-label="Stop recording"
+      >
+        <Icon content={StopCircle} size="lg" color={theme.colors.primary} />
+      </BaseButton>
```
I am not sure whether it would make it better or worse, but one thought I had here was to move this block to a function and re-use it like

```typescript
const createActionButton = (onClick, disabled, ariaLabel, iconContent, color) => {
  return (
    <BaseButton
        kind={BaseButtonKind.BORDERLESS_ICON}
        onClick={onClick}
        disabled={disabled}
        aria-label={ariaLabel}
      >
        <Icon content={iconContent} size="lg" color={color} />
      </BaseButton>
  )
}

if (recordPlugin?.isRecording()) {
  return createActionButton(stopRecording, disabled, "Stop recording", StopCircle, theme.colors.primary)
} 
else if (recordingUrl) {
  if (wavesurfer && wavesurfer.isPlaying()) {
    return createActionButton(onClickPlayPause, disabled, "Pause", Pause, theme.colors.fadedText60)  
  }

...
} 
```

It might make it easier to see on one glimpse that always the same Components (`<BaseButton><Icon/></BaseButton>`) are rendered.

But no strong opinion, just wanted to share my thought process 🙂

sfc-gh-nbellante replied:
if i have the time, happy to make this change

raethlein commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,332 @@
+  const [recordPlugin, setRecordPlugin] = useState<RecordPlugin | null>(null)
+  // to eventually show the user the available audio devices
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [availableAudioDevices, setAvailableAudioDevices] = useState<
```
is it possible to name the variable `_` instead of disabling `eslint` here as its possible in other languages? if yes, it might be a nicer convention

sfc-gh-nbellante replied:
tried that, but then our eslint naming convention rule complained about it not following naming conventions 😔

sfc-gh-nbellante replied:
to add on to this, i suspect that we will sooner rather than later need to add the ability to have some sort of drop down to select a mic. (and i'd rather not have to re-lookup how to do this 😅 )

raethlein commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,332 @@
+      return
+    }
+
+    const deviceId = activeAudioDeviceId
+    if (isNullOrUndefined(deviceId)) {
+      return
+    }
```
isn't this already checked via the check in [line 209](https://github.com/streamlit/streamlit/pull/9404/files#diff-a7eb8abbe0fa76f9716f9f399ee823661f1a6c761fa6065ef32204746da10ad4R209) `!activeAudioDeviceId`?

sfc-gh-nbellante replied:
yup, fixed

raethlein commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,332 @@
+  }, [theme, uploadTheFile])
+
+  useEffect(() => {
+    initializeWaveSurfer()
```
not sure whether this is a thing with waveSurfer, but is there an equivalent `deregisterPlugin` to the `registerPlugin` in [line 155](https://github.com/streamlit/streamlit/pull/9404/files#diff-a7eb8abbe0fa76f9716f9f399ee823661f1a6c761fa6065ef32204746da10ad4R155) and, if yes, would it make sense to call it in form of useEffect's returned callback when the component unmounts?

sfc-gh-nbellante replied:
hmm i didn't see anything like that in the docs. i think when destroy is called it all goes poof, but i can double check

sfc-gh-nbellante replied:
good intuition! looks like there is a way to deregister a plugin. https://github.com/katspaugh/wavesurfer.js/discussions/3238#discussioncomment-7206183

let me try that!

raethlein commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,332 @@
+      // we'll start keeping track of the playback time from that point onwards (until re-recording).
+      setShouldUpdatePlaybackTime(true)
+
+      // to get the pause button to show
+      forceRerender()
```
is this really needed here? calling `setShouldUpdatePlaybackTime(true)` in the line before already updates the state and should trigger a re-render, right?

sfc-gh-nbellante replied:
good point, that was added before the time thingy so i think this can be removed from here now 👍

raethlein commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,332 @@
+    widgetMgr,
+  ])
+
+  // Note: these can't be memoized due to the reliance on calling .isRecording() & .isPlaying()
```
though you could store `const isRecording = recordPlugin.isRecording()` here and use it in the next two checks if you believe that it is an expensive call

raethlein commented on frontend/lib/src/components/widgets/AudioInput/uploadFiles.ts:
```diff
@@ -0,0 +1,111 @@
+  error: Error
+}
+
+export const uploadFiles = async ({
```
I have not looked into this deeply, but could this function be shared with and re-used by the FileUploader?

sfc-gh-nbellante replied:
that was the goal at least! kind of ran out of time to complete that refactor so i'll add it as a backlog item

sfc-gh-nbellante replied:
added to tech debt db

raethlein commented on lib/streamlit/elements/widgets/audio_input.py:
```diff
@@ -0,0 +1,248 @@
+        *,  # keyword-only arguments:
+        disabled: bool = False,
+        label_visibility: LabelVisibility = "visible",
+    ) -> UploadedFile | None:
```
we unfortunately do not handle this consistently across widgets, but I want to highlight that using a Streamlit-defined type might not be the best UX because the developer sees the type, but has to click on it to get what it really means and we discourage developers to import it from here because we don't give any guarantees about our internal package structure.
I believe we should move all of our type definitions into a module `streamlit.types` or similar and make that an official part of our API (guarantee the existence) at some point. Until then, it might be nicer to duplicate the definition for the publicly exposed API here and write `Union[UploadedFile, DeletedFile, None]` (although also here two Streamlit-types are part of, so probably not worth it 😅)

No immediate action item for this PR, but wanted to share one inconsistency in our codebase right now so that we can all think about it 🙂

kajarenc replied:
@raethlein I definitely agree that types as part of Streamlit API is a great idea! 
We need to think carefully about exact types (especially for file-uploader, data-editor like widgets). But on the strategic level, it is something we should do at some point!

raethlein commented on lib/streamlit/elements/widgets/audio_input.py:
```diff
@@ -0,0 +1,248 @@
+        on_change: WidgetCallback | None = None,
+        args: WidgetArgs | None = None,
+        kwargs: WidgetKwargs | None = None,
+        *,  # keyword-only arguments:
```
super nit: I think `# keyword-only arguments:` is not needed as this is not an exotic Python construct here so I personally would leave it out because we probably don't comment it anywhere else in the codebase

raethlein commented on frontend/lib/src/components/widgets/AudioInput/ActionButton.tsx:
I believe we should have some JS unit tests as well like `ActionButton.test.tsx` and for the other files. What do you think?

sfc-gh-nbellante replied:
what would we gain by testing ActionButton with unit tests that we don't have covered from the e2e tests?

raethlein replied:
Our unit tests are orders of magnitudes faster than playwright tests, so during development you can run the tests quicker to see whether changes you made broke anything; and probably you also test more different input values than with e2e tests (also with respect to hooks etc). Additionally,if something is broken unit tests usually help to narrow down the issue faster because they are more fine granular vs the e2e tests that treat the widgets more like a blackbox. If something is broken and you only have e2e tests you have to spend time figuring out whether it's in the frontend or backend or both and then also what exactly.

sfc-gh-nbellante replied:
Solid argument, will add some unit tests for this

kajarenc committed:
remove obvious comment

sfc-gh-nbellante committed:
reverting toolbar changes

sfc-gh-nbellante committed:
cleanup the plugin too

sfc-gh-nbellante committed:
removing the forceRerender in one spot

sfc-gh-nbellante committed:
cleanup ActionButton

sfc-gh-nbellante committed:
adding snapshot

sfc-gh-nbellante committed:
adding kw

sfc-gh-nbellante committed:
adding snapshots

sfc-gh-nbellante committed:
getting themes in snapshot?

sfc-gh-nbellante committed:
themed snapshots

lukasmasuch commented on e2e_playwright/st_audio_input.py:
I think it would be good that the e2e test script covers all features: all parameters + usage in form/fragments, usage of callbacks. E.g. see the `st_feedback` test as an example of a pretty complete widget test: 

https://github.com/streamlit/streamlit/blob/2351b54056ac7f1c3500076f258e04dac7b65e57/e2e_playwright/st_feedback.py

Also, there are two basic element tests we added recently to all elements in the context of better CSS support:

`test_check_top_level_class` & `test_custom_css_class_via_key`:

https://github.com/streamlit/streamlit/blob/2351b54056ac7f1c3500076f258e04dac7b65e57/e2e_playwright/st_feedback_test.py#L132-L139

sfc-gh-nbellante replied:
I'm updating the tests but the fragment test seems to not be working, and I think that's a known issue given https://github.com/streamlit/streamlit/pull/9404#discussion_r1747474008

lukasmasuch replied:
I think the comment refers more to the complexity of implementing an interactive e2e test for these elements in `st_fragment_basic` and not to any issues with these elements in fragments. But I might be wrong.

@kajarenc are file uploader & camera input fully supported in fragment?

kajarenc replied:
@lukasmasuch: @raethlein and I manually tested the basic usage of those widgets (as well as audio_input with `st.fragment`). 

Basically nothing more serious than 
```
import streamlit as st


@st.fragment
def show_file_uploader():
    x = st.file_uploader("Upload photo:", type=["jpg", "png"])

    if x is not None:
        st.image(x)


show_file_uploader()
```
and it works as expected, although I don't know the nuances of how st. fragment works, so I can imagine that in some exotic scenarios, we could have bugs with file uploader widgets.

lukasmasuch replied:
@kajarenc Not urgent, but I think this could be eventually added as an e2e test case to the `st_file_uploader.py`.

sfc-gh-nbellante replied:
i added a test for fragments for file uploader!

sfc-gh-nbellante committed:
lint

sfc-gh-nbellante committed:
removing code

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/constants.ts:
```diff
@@ -0,0 +1,24 @@
+ * limitations under the License.
+ */
+
+export const WAVEFORM_HEIGHT = 68
```
nit: in the context of advanced theming, it might make sense to move this into the `sizes` properties in our theme:

https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/theme/primitives/sizes.ts

Based on the discussions, 68px (4.25rem) is supposed to be the target height for our "large" elements, so we could call this something like: `largeElementHeight`.

sfc-gh-nbellante replied:
okay, i did this, but needed to introduce a bit of [grossness](https://github.com/streamlit/streamlit/pull/9404/commits/7ac98897b74ace43ca2c6bdaaba480528841b8dd) in order to convert back to px for wavesurfer 😅

lukasmasuch replied:
Yeah, that's not super a bit suboptimal. But I think I will need this conversion logic soon for a couple of other places as part of the advanced theming project. Maybe you can already move this into theme utils as a utility method.

sfc-gh-nbellante committed:
cleaning up some bools

lukasmasuch commented on lib/streamlit/elements/widgets/audio_input.py:
```diff
@@ -0,0 +1,248 @@
+    from streamlit.delta_generator import DeltaGenerator
+
+if TYPE_CHECKING:
+    from streamlit.delta_generator import DeltaGenerator
```
Duplicate, see above

sfc-gh-nbellante committed:
a bit of cleanup and unit tests for AudioInputActionButton

sfc-gh-nbellante committed:
updating test

sfc-gh-nbellante committed:
removing exports

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,330 @@
+        // way wavesurfer is using these colors. Specifically, the progress color "tints" the wave color
+        // and the opacity of this wave color is not being respected, so to make things work for now hardcoding
+        // this color that looks "okay" on both themes.
+        waveColor: "#A5A5AA",
```
> this color that looks "okay" on both themes.

Does it also work ok for the current custom theme parameters? We are also in the process of preparing Streamlit to support advanced theming which might make it a requirement very soon to have this somehow be based on one of the theming properties. 

> the opacity of this wave color is not being respected

Does this refer to the `transparentize` usage of our theme properties? Maybe we can blend the color with the background color to get the correct color. Unfortunately, our color library (color2k) does not provide a blend method, but it should be pretty easy to do, e.g.:

```
import { parseToRgba, toHex } from "color2k"

export function blend(color: string, background: string | undefined): string {
    if (background === undefined) return color;
    const [r, g, b, a] = parseToRgba(color);
    if (a === 1) return color;
    const [br, bg, bb, ba] = parseToRgba(background);
    const ao = a + ba * (1 - a);
    // (xaA + xaB·(1−aA))/aR
    const ro = (a * r + ba * br * (1 - a)) / ao;
    const go = (a * g + ba * bg * (1 - a)) / ao;
    const bo = (a * b + ba * bb * (1 - a)) / ao;
    return toHex(`rgba(${ro}, ${go}, ${bo}, ${ao})`);
}
```

> Specifically, the progress color "tints" the wave color

Do you know the reason for the tint? Does it mean that the actual color needs to be slightly different to actually get the color we want?

sfc-gh-nbellante replied:
Okay, I did some more looking into this. Here's the problem:
fadedText40 -- the faded part comes from the fact that transparency is used
progressWaveColor -- the wavesurfer property is a _tint_ not a color that just replaces it

the lower opacity on fadedText40 carries over to the tint making it not possible to tint it black how we want. (even if i specify "#000")

so in order to fix this, we need a color that has 100% opacity that comes out to be the same color as fadedText40 does on the background its on. does that make sense?

sfc-gh-nbellante replied:
so are you suggesting taking the fadedText40 and using that blend function to blend it with the background and then setting that resulting color as the waveColor?

lukasmasuch replied:
Yep, that sounds like using `waveColor = blend(fadedText40, background)` could work and might be worth a try. Btw. the blend code above is from [glide-data-grid](https://github.com/glideapps/glide-data-grid) (the underlying data grid library of Streamlit) where it is used a lot for similar aspects. But not 100% sure if it works correctly for the case here. If it doesn't work correctly, I can also take another look on this later with the advanced theming project.

sfc-gh-nbellante replied:
whew... for a second i thought you just casually pulled that algo out of thin air 😅 . let me try it and see what happens

sfc-gh-nbellante replied:
so i got this to work, but only by putting Math.round(...) around the ro, bo, and go lines because it was trying to convert floats in the rgba to hex and struggling. i think this color should be close enough that it's good enough for now? wdyt?

lukasmasuch replied:
Yep, I think this sounds reasonable 👍 Thanks for taking another look!

sfc-gh-nbellante committed:
lint

sfc-gh-nbellante committed:
fix test

sfc-gh-nbellante committed:
switching to hook instead of withTheme

sfc-gh-nbellante committed:
updating testIds

sfc-gh-nbellante committed:
adding css class and moving testId

sfc-gh-nbellante committed:
refactor height to theme

sfc-gh-nbellante committed:
removing duplicate

sfc-gh-nbellante committed:
enhancing tests

sfc-gh-nbellante committed:
removing the hardcoded color, thanks lukas!

sfc-gh-nbellante committed:
fixing lint

sfc-gh-nbellante committed:
updating snapshots

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,324 @@
+      uploadTheFile(file)
+
+      ws.setOptions({
+        // This color is hardcoded because I've spent an hour trying to get it to work
+        // with the theme and I'm giving up for now. The complications arise due to the
+        // way wavesurfer is using these colors. Specifically, the progress color "tints" the wave color
+        // and the opacity of this wave color is not being respected, so to make things work for now hardcoding
+        // this color that looks "okay" on both themes.
```
nit: This comment can be removed  :) Maybe add a comment on why we blend the color and not directly use `fadedText40` instead

kajarenc committed:
make all st.audio_input arguments keyword only (except label).

sfc-gh-nbellante committed:
replacing comment with useful one

sfc-gh-nbellante committed:
make fragments work

sfc-gh-nbellante committed:
adding hover color

sfc-gh-nbellante committed:
lint

sfc-gh-nbellante committed:
fix py

sfc-gh-nbellante committed:
adjusting buttons

sfc-gh-nbellante committed:
updating jest tests

sfc-gh-nbellante committed:
updating bar radius

sfc-gh-nbellante committed:
fix lint

sfc-gh-nbellante committed:
updating snapshots

lukasmasuch commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -0,0 +1,138 @@
+def test_snapshots(themed_app: Page, assert_snapshot: ImageCompareFunction):
+    audio_input_element = themed_app.get_by_test_id("stAudioInput").first
+
+    assert_snapshot(audio_input_element, name="st_audio_input_default")
```
super nit: the screenshot pattern we mostly use would be `st_audio-input_default` (`<component>-<screenshot description>`). But no need to change...

sfc-gh-nbellante replied:
wait, so then shouldn't it be `st_audio_input-default`?

lukasmasuch replied:
ahh, yeah... sorry :)

lukasmasuch commented on e2e_playwright/st_audio_input.py:
It would be great if the e2e test covers all user-facing features. Missing:
- Usage of callback (see [this test](https://github.com/streamlit/streamlit/blob/671730333c6caebbe167cc25e383ca1263cbeb3f/e2e_playwright/st_slider.py#L74-L87))
- Usage of `disabled` (via screenshot test)
- Usage of `label_visibility` (via screenshot test)
- Usage of `help` (via `app_utils.expect_help_tooltip`). This seems to be only partly implemented; it is part of _audio_input and the proto, but not exposed via the public API.
- Test that it keeps state after it gets unmounted (see [this example](https://github.com/streamlit/streamlit/blob/671730333c6caebbe167cc25e383ca1263cbeb3f/e2e_playwright/st_feedback.py#L48-L56))

sfc-gh-nbellante replied:
added all 5 of these, thanks for clearly laying these examples out

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,342 @@
+  )
+}
+
+export default AudioInput
```
nit: @raethlein recently figured out that it can drop unnecessary renderings a lot of this is put into a memo here.

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/uploadFiles.ts:
This hook looks quite generic. Can we reuse it in other widgets with upload file support (e.g. in the upcoming file upload support for chat input)? If so, maybe we can we move this to `lib/src/hooks`?

sfc-gh-nbellante replied:
that's the goal! good call, let me move it to a more generic spot

lukasmasuch commented on e2e_playwright/__snapshots__/linux/st_audio_input_test/st_audio_input_default[light_theme-webkit].png:
What's the reason why this looks different compared to the other snapshots? Are the permissions working differently in webkit?

sfc-gh-nbellante replied:
yeah, in webkit it says unknown permission "microphone"

lukasmasuch commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -0,0 +1,138 @@
+
+
+@pytest.mark.only_browser("chromium")
+def test_feedback_works_with_fragments(app: Page):
```
*audio_input

lukasmasuch commented on e2e_playwright/st_audio_input.py:
```diff
@@ -0,0 +1,48 @@
+@st.experimental_fragment()
+def test_file_fragment():
+    file_uploader_in_fragment = st.file_uploader(label="file uploader")
+    st.write("File uploader in Fragment:", file_uploader_in_fragment)
```
I think this test could be better moved to `st_file_uploader.py`?

lukasmasuch commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -0,0 +1,138 @@
+
+    fragment_audio_input = app.get_by_test_id("stAudioInput").nth(2)
+    fragment_audio_input.get_by_role("button", name="Record").click()
+    time.sleep(1)
```
What is the reason for the `sleep` here? Is it to wait for some recording? I think in playwright its better to use `app.wait_for_timeout()` instead of the Python sleep methods: https://playwright.dev/python/docs/library#timesleep-leads-to-outdated-state

sfc-gh-nbellante replied:
yes, to wait for recording. oh okay, i'll switch

lukasmasuch commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -0,0 +1,138 @@
+    stop_button = audio_input.get_by_role("button", name="Stop recording").first
+    expect(stop_button).to_be_visible()
+
+    time.sleep(2)
```
See the comment above about `time.sleep`

lukasmasuch commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -0,0 +1,138 @@
+
+    time.sleep(2)
+
+    stop_button.click()
```
Maybe also add a `wait_for_app_run` here and check via the written value that the value was actually received by the frontend as done in the form and fragment case.

lukasmasuch commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -0,0 +1,138 @@
+    clear_button = audio_input.get_by_role("button", name="Clear recording").first
+    expect(clear_button).to_be_visible()
+
+    clear_button.click()
```
Clearing should also trigger a rerun and clear the value on the backend? Thats something that could also be verified here.

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
I think it would be good to have a very basic `renders without crashing` test that we also have for most other elements in `AudioInput.test.tsx`. Maybe also a few more tests on behaviour, but for me it is fine to keep the RTL tests simple as long as the e2e test cover all user-facing aspects.

sfc-gh-nbellante replied:
idk if i'm doing this wrong... but it seems like in order to get rtl tests working for audio input i need to mock A LOT of stuff and at that point it makes me wonder where the value is in adding this test. knowing this, do you still feel strongly we need this test?

sfc-gh-nbellante replied:
specifically in the media devices, what a media stream is, the methods attached to them etc

lukasmasuch replied:
Oh I see... so it would run into errors if you just mock all the input props (element, uploadClient, widgetMgr, fragmentId) because of some missing media device aspects?

lukasmasuch reviewed (commented):
Great implementation & feature 🚀 it looks almost ready for me, I have a couple comments on testing and some observations here: 

The responsiveness of the widget seems to be quite good. The case where it doesn't work well is when it is missing the microphone permission (maybe the `StyledWaveformContainerDiv` needs an `overflow: hidden`?):
 
<img width="291" alt="image" src="https://github.com/user-attachments/assets/6ac2b189-a744-4837-9e8c-4c8dc47a2b97">

Another observation: clicking on start recording - if there is already an old recording - triggers a rerun. This rerun seems to be a bit unnecessary. Is this expected? 

https://github.com/user-attachments/assets/d16465a5-4e82-45f5-8609-ceb9b8b64aad

A strange observation: changing the theme via the menu seems to consistently cause some trouble with the placing of the waves: 

<img width="743" alt="image" src="https://github.com/user-attachments/assets/a3cd8221-23c3-49c0-a48b-b61abaeb4d39">

sfc-gh-nbellante committed:
fix name

sfc-gh-nbellante committed:
moving test and cleaning up test file i think? or breaking a bunch of stuff... we'll find out

sfc-gh-nbellante committed:
fixing time.sleeps

sfc-gh-nbellante commented:
> Another observation: clicking on start recording - if there is already an old recording - triggers a rerun. This rerun seems to be a bit unnecessary. Is this expected?

@lukasmasuch the reason the rerun is happening is because its deleting the old file before recording the new. perhaps we should not do that to avoid this rerun?

sfc-gh-nbellante committed:
moving the util to more generic spot

sfc-gh-nbellante committed:
use memo on the components

sfc-gh-nbellante committed:
check that the value roundtripped on basic test

sfc-gh-nbellante committed:
add overflow hidden to StyledWaveformContainerDiv

sfc-gh-nbellante committed:
make sure the clear round tripped

lukasmasuch commented:
> the reason the rerun is happening is because its deleting the old file before recording the new. perhaps we should not do that to avoid this rerun?

Hmm, yeah it looks like we don't need to delete the file / rerun up until the point where the new recording is available (on stop recording)? Two potential options:
- on start recording: delete file but don't call `setFileUploaderStateValue` 
- or delete the old recording file only on `stopRecording` before uploading the new data (maybe the safer option)

Clicking the clear button might be the only case where we want to delete the file & `setFileUploaderStateValue` to empty.

lukasmasuch commented on e2e_playwright/__snapshots__/linux/st_audio_input_test/st_audio_input_default[light_theme-chromium].png:
Regarding `help` label: Wouldn't it be possible to add the help tooltip right to the widget label like we do for all other widgets?

kajarenc replied:
@lukasmasuch My current understanding is that from the product perspective, we want to release it without help tooltip for now, and it later https://snowflake.slack.com/archives/C07B3C1NVMF/p1725428454820699?thread_ts=1725409523.392919&cid=C07B3C1NVMF

But let's double-check it with the product; if it's okay, I can add help to the public API, too.

sfc-gh-nbellante replied:
convo happened in slack but tldr is we implemented the help right next to the label

sfc-gh-nbellante committed:
fix file uploader test

kajarenc committed:
add help parameter back to the public API

sfc-gh-nbellante committed:
clear recording URL on record end instead of start

sfc-gh-nbellante commented:
> or delete the old recording file only on stopRecording before uploading the new data (maybe the safer option)

i've switched it to follow this suggestion

sfc-gh-nbellante committed:
adding some more e2e tests

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,342 @@
+  fragmentId?: string
+}
+
+const AudioInput: React.FC<Props> = ({
```
Oh, I think one case that needs to be handled is the `disabled` state via the `disabled` prop that gets passed in from `widgetProps` in `ElementNodeRenderer`. I think you can just get it from the props here and use it instead of `element.disabled` everywhere.

Context: there are some niche situations where the frontend disables all widgets independent of what is configured by the user. This happens in a situation like when the app loses the connection.

sfc-gh-nbellante replied:
good catch!

sfc-gh-nbellante committed:
adding callback test

sfc-gh-nbellante committed:
updating snapshot names

sfc-gh-nbellante committed:
lint

sfc-gh-nbellante committed:
fixing disabled

sfc-gh-nbellante committed:
fix test

kajarenc committed:
add audio_input to list of mocked widget elements

sfc-gh-nbellante committed:
fixing tests

sfc-gh-nbellante committed:
fixing the bug introduced when trying to eliminate the rerun on rerecord and also fixing the theming bug

sfc-gh-nbellante committed:
updating snapshots

sfc-gh-nbellante committed:
adding dep

sfc-gh-nbellante committed:
mac snaps

sfc-gh-nbellante committed:
adding help, fixing bug with toolbar not showing because overflow hidden

sfc-gh-nbellante committed:
use the util

sfc-gh-nbellante committed:
fixing bare execution tests

sfc-gh-nbellante committed:
LINNNNNNNNNNNNT

sfc-gh-nbellante committed:
not doing this test due to the level of mocking necessary

lukasmasuch commented on e2e_playwright/__snapshots__/linux/st_audio_input_test/st_audio_input-disabled[light_theme-chromium].png:
The widget label for `disabled` state might need a different grey; see this text input as a comparison: 

<img width="590" alt="image" src="https://github.com/user-attachments/assets/5a121e5c-3774-4fd5-9610-7b7833069204">

sfc-gh-nbellante replied:
this was due to the missing disabled that you pointed out in the other comment

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -0,0 +1,393 @@
+    >
+      <WidgetLabel
+        label={element.label}
+        disabled={hasNoMicPermissions}
```
`disabled={disabled || hasNoMicPermissions}`

lukasmasuch commented on __snapshots__/darwin/st_audio_input_test/st_audio_input-default[dark_theme-chromium].png:
nit: I think the darwin snapshots here might have been accidentally pushed

sfc-gh-nbellante replied:
should that folder be git ignored?

sfc-gh-nbellante replied:
nvm it's because it dumped the screenshots in the wrong parent folder

lukasmasuch commented on frontend/lib/package.json:
```diff
@@ -60,7 +61,7 @@
     "decamelize": "^6.0.0",
     "deck.gl": "^8.9.36",
-    "dompurify": "^3.1.3",
+    "dompurify": "^3.0.9",
```
nit: this might need to be changed back to `3.1.3` (from a recent security dependabot update)

kajarenc committed:
add small changes to `compute_and_register_element_id` usage

sfc-gh-nbellante committed:
updating disabled prop

sfc-gh-nbellante committed:
fix package

sfc-gh-nbellante committed:
updating disabled screenshots

sfc-gh-nbellante committed:
removing snapshots

sfc-gh-nbellante committed:
adding comments to e2e tests

sfc-gh-nbellante committed:
cleaning up streamlit test code

sfc-gh-nbellante committed:
removing console logs and fixing equality check

sfc-gh-nbellante committed:
updating snapshots?

lukasmasuch commented on e2e_playwright/st_audio_input.py:
I tried out the latest e2e script and it looks pretty complete. Thanks! :) 
One aspect that is not fully working yet is the unmounting / remounting:

https://github.com/user-attachments/assets/cf327da6-9112-457b-b63c-e34c1ffe43c1

While the backend state is kept, the frontend component itself loses its state -> this creates issues since the state shown in the frontend component is out of sync with the state of the backend. See the file uploader as an example which keeps the state of the uploaded file. 

The reconstruction of the state from the widget manager happens here in the file uploader: 

https://github.com/streamlit/streamlit/blob/67aaff5e25bf73cf0a7dddcfd886e0f083fde6bb/frontend/lib/src/components/widgets/FileUploader/FileUploader.tsx#L186

You can also find similar logic in most of the other widgets.

sfc-gh-nbellante replied:
Hmmm so just to confirm, we would be okay with just preserving the waveform, but not the play/pause/seek location state, right?

sfc-gh-nbellante replied:
also seems like i should add some sort of check into these tests that looks at the canvas and determines if a waveform is drawn or something of that nature so that we don't get false positive test results as we currently are getting with this test.

lukasmasuch replied:
In the optimal case, the user wouldn't notice that the elements got recreated. But reconstructing the full state can get quite tricky for complex components. Therefore, I think it is more of a balance of reconstructing the most important aspects. E.g. if the user has paused the playback, its not something we need to reconstruct so that its still paused at the exact position. Thought, being able to play the audio would be kind of nice. 

You can store and retrieve additional frontend-only state into the widget manager by using `widgetMgr.setElementState(element.id, "<key>", <object>)`. E.g we are using this in more complex components like `PlotlyChart`.

sfc-gh-nbellante replied:
ah okay, thanks for pointing this out!

lukasmasuch replied:
> seems like i should add some sort of check into these tests that looks at the canvas and determines if a waveform is drawn or something of that nature so that we don't get false positive test results as we currently are getting with this test.

Yep, that would be good 👍

sfc-gh-nbellante replied:
idk if there's a better way, but since we're rendering the placeholder with a background repeat, the only canvas is the wavesurfer canvas so i made the test check to see if the wavesurfer canvas is visible. granted, this doesn't ensure the waveform is the right waveform.. but i can't think of a better way to check that reliably

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/styled-components.ts:
```diff
@@ -0,0 +1,132 @@
+      padding: theme.spacing.threeXS,
+      color: theme.colors.fadedText40,
+    },
+    "&:hover > button, &:focus > button": {
+      color: theme.colors.primary,
+    },
+    "&:disabled > button": {
+      color: theme.colors.fadedText40,
+    },
```
Something here might not be working 100% with disabled. It is still possible sometimes trigger the hover effect when disabled:

https://github.com/user-attachments/assets/786132ed-1270-45fb-9ba1-f42690ef6797

I believe the `&:disabled` isn't really doing anything here since the span doesn't have a disabled set as an attribute. It might  be better to directly modify the hover/focus state of the button. 

```
    "& > button:hover, & > button:focus": {
      color: theme.colors.primary,
    },
```

In that case, you don't have to do something with disabled state since it is already handled by the borderless button. What wouldn't hurt is to add a `cursor: "not-allowed",` to the disabled in styled-components of BaseButton.

sfc-gh-nbellante replied:
nice catch! 🦅 👁️

sfc-gh-nbellante committed:
break the tests (or rather make them fail because they should be failing)

sfc-gh-nbellante committed:
make it retain state after unmount

sfc-gh-nbellante committed:
fix hover states

sfc-gh-nbellante committed:
linttttttttttttttt

sfc-gh-nbellante committed:
refactor: update audio input tests and implement useWidgetManagerElementState hook

sfc-gh-nbellante committed:
only request the mic perm when the user starts to record

lukasmasuch commented on frontend/lib/src/components/shared/BaseButton/styled-components.ts:
```diff
@@ -251,6 +253,7 @@ export const StyledIconButton = styled(
       backgroundColor: theme.colors.lightGray,
       borderColor: theme.colors.transparent,
       color: theme.colors.gray,
+      cursor: "not-allowed",
```
nit: i might need to add it to the **borderless** icon button as well :)

sfc-gh-nbellante committed:
updating snapshots

sfc-gh-nbellante committed:
cleanup

lukasmasuch commented on frontend/lib/src/hooks/useWidgetManagerElementState.tsx:
```diff
@@ -0,0 +1,58 @@
+  notNullOrUndefined,
+} from "@streamlit/lib/src/util/utils"
+
+const useWidgetManagerElementState = <T,>({
```
nice addition 🥳 

nit: maybe add a small docstring comment that explains what the hook can be used for.

lukasmasuch commented on frontend/lib/src/components/widgets/AudioInput/uploadFiles.ts:
There is also a version of `uploadFiles.ts` that looks pretty similar / the same in `frontend/lib/src/util/uploadFiles.ts`. Do we need both?

sfc-gh-nbellante replied:
huh strange.. i thought i moved this but perhaps i copied and forgot to delete. but yeah this can be deleted now

sfc-gh-nbellante committed:
updating tests

lukasmasuch reviewed (approved):
LGTM 🎤🥳  A few remaining comments

lukasmasuch commented:
Oh, there is something weird going on. Not sure if it has to do with my setup. But it seems that the first audio input doesn't correctly transfer the audio to the backend:

https://github.com/user-attachments/assets/b22404a6-b054-44f8-9ee3-6b52a66dac75

Can you also reproduce this?

sfc-gh-nbellante commented:
> Oh, there is something weird going on. Not sure if it has to do with my setup. But it seems that the first audio input doesn't correctly transfer the audio to the backend:
> 
>  Screen.Recording.2024-09-19.at.21.55.16.mov 
> Can you also reproduce this?

I cannot reproduce this.

![Kapture 2024-09-19 at 13 03 06](https://github.com/user-attachments/assets/aa2429cd-de4f-415a-9651-30e967baf341)

sfc-gh-nbellante committed:
adding comment to useWidgetManagerElementState

sfc-gh-nbellante committed:
delete this file

kajarenc committed:
Rename `st.audio_input` to `st.experimental_audio_input` (#9531)

## Describe your changes
Rename `st.audio_input` to `st.experimental_audio_input` 

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
adding safari blockade

sfc-gh-nbellante committed:
lint

sfc-gh-nbellante committed:
updating test and snapshots

sfc-gh-nbellante committed:
couple more snapshot updates

sfc-gh-nbellante committed:
updates based on feedback + some more tests

sfc-gh-kmcgrady reviewed (approved):
Protobuf is backwards-compatible

sfc-gh-braethlein commented on proto/streamlit/proto/AudioInput.proto:
```diff
@@ -0,0 +1,33 @@
+import "streamlit/proto/LabelVisibilityMessage.proto";
+
+
+message AudioInput {
+  string id = 1;
+  string label = 2;
+  string help = 3;
+  string form_id = 4;
+  bool disabled = 5;
+  LabelVisibilityMessage label_visibility = 6;
```
I think this is another aspect we should codify: whether or not to use `optional` I think we use it pretty inconsistently 😅
